Add datashuttle blog (#98)
neuroinformatics-unit/neuroinformatics-unit.github.io: docs/source/blog/datashuttle.md

---
blogpost: true

> date: 25 April 2024

Reviewer: BTW these dates can be set in the future if needed.

author: Joe Ziminski, Niko Sirmpilatze
location: London, UK
category: Blog
language: English
image: 1
---

# Managing neuroscience projects with **datashuttle**
*Create, validate and transfer standardised project folders*

```{image} /_static/blog_images/datashuttle/datashuttle-overview-dark.png
:align: center
:class: only-dark
:width: 650px
```
```{image} /_static/blog_images/datashuttle/datashuttle-overview-light.png
:align: center
:class: only-light
:width: 650px
```
<br>

> Maintaining a well-organised neuroscience project is hard. 

Reviewer: ```suggestion
**Maintaining a well-organised neuroscience project is hard.**
```

Although everyone can appreciate the benefits of a tidy project
folder, the practicalities of running an experiment often gets 
in the way. Folder organisation 
is low on the list of priorities when it
comes to acquisition sessions spent managing complex systems 
and experimental animals.

However, the cost of small mistakes during data acquisition can be high.
One misplaced character may mean sessions are missed by analysis 

> scripts or subject identifications duplicated.

Reviewer 2: I feel there's something syntactically wrong wit this sentence, some alternatives:

1. One misplaced character may mean sessions are missed by analysis 
scripts or subject IDs are duplicated.

2. One misplaced character may lead to subjects with duplicate IDs or to sessions being skipped by analysis scripts.

Author: How about `subject identifiers`? I want to avoid ID if possible just to avoid abbreviations.

Reviewer 2: identifiers is fine.

> The best protection against such errors is automating the process

Reviewer 2: This sentence is a bit long-winded, because of the dashes etc. Consider rewording or breaking up.

Author: How about:

```
The best protection against such errors is automating the process
through acquisition scripts—resulting in hours spent writing 
data-management code entirely unrelated to the central research goals.
```
I am also not sure whether so say `your` research goals rather than `the` to make it more personal. But it might be too much.

Reviewer 2: I'd use "your" in this case, but more consistency throughout is more important.

through acquisition scripts—resulting in hours spent writing code 
managing the naming and transfer of folders—a task entirely 
unrelated to the central research goals.

In our previous blog, we highlighted the benefits of data standardisation 
for systems neuroscience, introducing the 
[NeuroBlueprint](https://neuroblueprint.neuroinformatics.dev/) 
specification. 
An immediate benefit of a widely-used standard is that the entire community
can contribute to shared tools for project management.
This means individual researchers don't waste time
writing data-management code. 

> In this blog post we introduce 

Reviewer 2: Earlier you say "In our previous blog", while here you say "In this blog post". I think it should be "blog post" in both cases.

[**datashuttle**](https://datashuttle.neuroinformatics.dev/)—a 

> tool for the automated creation, 

Reviewer 2: I like this tagline.

validation and transfer of projects organised to 

> the **NeuroBlueprint** standard. **datashutte** aims to

Reviewer: ```suggestion
the **NeuroBlueprint** standard. **datashuttle** aims to
```

drop into existing acquisition pipelines, reducing errors
associated with manual folder creation and removing the need

> to write your down data-management code.

Reviewer: What does this mean?

Author: Hmm I think this is an example of the below `awkwardness of this blogpost is it is trying to cater to both those who do data management manually as well as automated.`

It is basically saying three things at once:
- datashuttle should just drop into your existing pipeline at the stage of creating folders, and transferring them at the end of the session
- if you are creating your folders manually, it will help reduce typographical errors
- if you are writing your down data manangement code, it means you no longer have to.

Reviewer: This sentence makes sense, I think I must have commented on the wrong line!

Reviewer 2: I like how you have made the point here in these 3 bullet points. Maybe that's how we should argue it?

Author: >This sentence makes sense, I think I must have commented on the wrong line!

Oh I think it must be the "**down** data-management code." instead of own 😅

Author: how about:

```
And that really is all there is to **datashuttle**, which aims to, 
drop into your existing acquisition pipelines whether they be manual
or automated:
- the **graphical interface** can be used instead of manual folder creation, 
reducing the risk of typographical errors
- alternatively, the **Python API** can be used directly in automated acquisition 
pipelines, removing the need to write you own data-management code

Below we will give an overview of **datashttle**'s key folder creation
and transfer features.
```

(also introducing the two ways of running it here).

Reviewer: Good to introduce the two ways of running it, but that first sentence is kinda weird (and has dodgy grammar). 

How about:

>*datashuttle**, aims to drop into your existing acquisition pipelines whether they be manual
or automated, with two ways to use it:
>- The **graphical user interface** can be used instead of manual folder creation, 
reducing the risk of typographical errors
>- The **Python API** can be used directly in automated acquisition 
pipelines, removing the need to write you own data-management code

>Below we will give an overview of **datashttle**'s key folder creation
and transfer features.

Author: Great thanks!

> Below we give a whistlestop tour of **datashuttles** and it's key

Reviewer: ```suggestion
Below we give a whistlestop tour of **datashuttle** and it's key
```

Reviewer 2: It's the first time a hear the term "whistlestop". Maybe replace with "quick" or something else?

features. More information is available at the 

> [datashutte](https://datashuttle.neuroinformatics.dev/) 

Reviewer: ```suggestion
[datashuttle](https://datashuttle.neuroinformatics.dev/) 
```

website.

## How **datashuttle** is used in an experiment

```{image} /_static/blog_images/datashuttle/tutorial-1-example-file-tree-dark.png
:align: center
:class: only-dark
:width: 400px
```
```{image} /_static/blog_images/datashuttle/tutorial-1-example-file-tree-light.png
:align: center
:class: only-light
:width: 400px
```
<br>

> Imagine that you are starting a new experiment and have the first

Reviewer 2: "during which you acquire both behavioural (`behav`) and electrophysiological (`ephys`) data" ?

Reviewer 2: also, this sentence should probably go before the image showing the data.

acquisition session, or behaviour (`behav`) 
and electrophysiological (`ephys`) data. 

The first thing typically done prior to acquiring data is to

> create the folders that the data will be stored. 

Reviewer: ```suggestion
create the folders that the data will be stored in. 
```

**datashuttle** can be used to quickly create the standardised
project folders in which to store the acquired data. 
Using **datashuttle** as apposed to manual creation or using custom

> scripts is that it ensures no formatting errors creep in.

Reviewer 2: ```suggestion
scripts ensures that no formatting errors creep in.
```

> At the end of the session once data is acquired, 

Reviewer 2: ```suggestion
At the end of the data acquisition session,
```

**datashuttle** transfers the newly acquired data 
to a central storage machine to be backed up.

Later on in the experiment, you may want to transfer only a subset

> of data from the central machine to an analysis machine. You may 

Reviewer 2: "to a computes used for data analysis"? 
"analysis machine" may not be immediately obvious without explanation.

want to pilot a behavioural data analysis, and grab

> for example only the behavioural 'test' session for

Reviewer 2: I think we can skip mentioning 'test' sessions to avoid confusion.
Fetching only behavioural data from the first 5 subjects is a clear enough example.

the first 5 subjects. **datashuttle** allows flexible custom transfers

> easily, meaning you don't have to drag and drop these data manually or

Reviewer 2: "datashuttle allows for easy and flexible data transfers"?
Saying both "flexible" and "custom" feels redundant.

> write a custom script.

Reviewer: Would it be worth rephrasing this as:

When organising data manually you would
* A (which can cause error X)
* B (which can cause error Y)
* C (which can cause error Z)

With datashuttle you can:
*
*

Author: I like the idea of laying it out like that. One issue that is a general awkwardness of this blogpost is it is trying to cater to both those who do data management manually as well as automated. The only worry about emphasising the (many) potential errors caused by manual too much is people who are using scripts think 'well, that's not a problem for me' and ignore it.

Reviewer: Oh yeah, I hadn't thought about our two "target markets!

And that really is all there is to **datashuttle**, a tool to drop into acquisition
workflows to make folder creation and transfer more convenient and ensure standardisation.
Below we will give a brief tour of these key **datashuttle** features.

> ## Creating folders with live-validation

Reviewer: Maybe before this give very brief details about installation, and how the GUI runs? Just to illustrate that it's pretty simple and runs anywhere?

Reviewer: Also state early that there's a GUI and Python API that can be used interoperably?

Author: > Maybe before this give very brief details about installation, and how the GUI runs? Just to illustrate that it's pretty simple and runs anywhere?

I'm in two minds about this, it is nice to give a quick overview of getting setup. The issue is there is a function to run `make_config_file` which requires some understanding of local vs. central path concept, which it might be a bit too much to get into here. I don't want to give the impression you can just install and run `datashuttle launch` then start creating files as it might be confusing when people actually go to use it. At the same time I dont really want to discuss the setup as it is TMI🤔 so currently just ignore entirely.

Reviewer: Maybe describe it at a high level? Mention install via conda or PyPi, and runs on any system (windows, Linux, macOS), including via remote terminals?

Reviewer: Just to get ahead of anyone thinking "The GUI looks nice, but I usually work remotely via ssh, so this won't work me", or something.

Author: I added the below to the start of the 'How datashuttle is used in an experiment', let me know how it reads / what you think of its placement

**datashuttle** runs on Windows, macOS or Linux and is
[easy to install](https://datashuttle.neuroinformatics.dev/pages/how_tos/install.html)
through 
[Conda Forge](https://conda-forge.org/)
or 
[PyPi](https://pypi.org/). **datashuttle** can be used
from within Python code (using the *Python API*) or through a graphical 
interface that works in any system terminal.

Reviewer: LGTM

> To create folders manually through the graphical interface with **datashuttle**, 

Reviewer 2: Slightly grammatically awkward.
I'd rephrase as "Creating folders through **datashuttle**'s graphical interface is as simple as..."

it is as simple as entering the subject and session name in the folder and clicking 'create'.

```{image} /_static/blog_images/datashuttle/create-folders-example-dark.png
:align: center
:class: only-dark
:width: 650px
```
```{image} /_static/blog_images/datashuttle/create-folders-example-light.png
:align: center
:class: only-light
:width: 650px
```
<br>

There are a number of convenience shortcuts to ensure you only need to enter 
your custom information. For example, the convenience tags 

> (`@DATE@`, `@TIME@`, `@DATETIME@`) will fill the created folder 

Reviewer: If a naive reader doesn't know what a "convenience tag" is, they may not see them as being very convenient!

:align: center
:class: only-dark
:width: 650px
```
```{image} /_static/blog_images/datashuttle/how-to-transfer-custom-light.png
:align: center
:class: only-light
:width: 650px
```
<br>

and in code, it looks similar:

```python
from datashuttle import DataShuttle

project = DataShuttle("my_first_project")

project.transfer_custom(
    "rawdata", "all_sub", "ses-001_@*@", "behav"
)
```

## Logging
A final feature of **datashuttle** is logging—whenever a folder is created or
data transferred, full details are saved in the logs. This ensures
a full history of the project is available at any time

***TODO ADD AN IMAGE OF LOGS IN GRAPHICAL INTERFACE**

## Getting started with **datashuttle**

We have given a brief tour of **datashuttle**'s key features,
but full details on getting started can be found at our 
[website](https://datashuttle.neuroinformatics.dev/) and
[getting started tutorial](https://datashuttle.neuroinformatics.dev/pages/tutorials/getting_started.html).

We are very keen to get feedback on **datashuttle**. 
Standardisation is incredibly useful, but it should not come at the 
expense of convenience. **datashuttle** aims to make managing your project easier than 

> it currently is—if it is not, we want to hear how it can be improved. 

Reviewer 2: Nice!

**datashuttle** allows you to transfer project folders
at the click of a button (graphical interface) or with a 
single function call (Python API).

Later on in the experiment, you may want to transfer only a subset
of data from the central machine to separate computer for analysis. For example,
you may  want to pilot an animal tracking pipeline, grabbing
only the behavioural data for the first 5 subjects. 
**datashuttle** allows flexible custom transfers, 
meaning you don't have to drag and drop these data manually or
write a custom script.

**datashuttle** aims to drop into your existing acquisition pipelines whether 
they be manual or automated, with two ways to use it:

- The **graphical user interface** can be used instead of manual folder creation
or transfer, reducing the risk of errors
- The **Python API** can be used directly in automated acquisition
pipelines, removing the need to write you own data-management code

Below we will give an overview of **datashuttle**'s key folder creation
and transfer features.

## Creating folders with live-validation

Creating folders through **datashuttle**'s graphical interface is as simple as

> entering the subject and session names, selecting the datatype and clicking 'create':

Reviewer: For consistency with below.
```suggestion
entering the subject and session names, selecting the datatype and clicking `Create Folders`
```

```{image} /_static/blog_images/datashuttle/create-folders-example-dark.png
:align: center
:class: only-dark
:width: 650px
```
```{image} /_static/blog_images/datashuttle/create-folders-example-light.png
:align: center
:class: only-light
:width: 650px
```
<br>

Live-validation of inputs as you type ensures
formatting errors don't creep into the project:

```{image} /_static/blog_images/datashuttle/validation-bad-dark.png
:align: center
:class: only-dark
:width: 500px
```
```{image} /_static/blog_images/datashuttle/validation-bad-light.png
:align: center
:class: only-light
:width: 500px
```
<br>

There are a number of shortcuts to ensure only

> information relevant to your specific project must be typed. 

Reviewer 2: I also coudn't parse this sentence. I'll rewrite it as:
"There are a number of shortcuts to reduce the amount of manual typing."

For example, the tags (`@DATE@`, `@TIME@`, `@DATETIME@`) will 
fill the created folder  with the date / time / datetime. Double-clicking
an input will suggest the next subject or session.
A full list of shortcuts (e.g. automatic suggestion of next 
subject / session names) are available in the 
[documentation.](https://datashuttle.neuroinformatics.dev/pages/how_tos/create-folders.html#creating-project-folders)

Folders can be created in an equivalent way through the Python API:

> ```python
> from datashuttle import DataShuttle             # import datashuttle
> 
> project = DataShuttle("my_first_project")       # initialise the project
> 
> created_folder_paths = project.create_folders(  # create the folders
>     "sub-001", "ses-001_@DATE@", ["behav", "funcimg"]
> )
> ```

Reviewer: I don't think the comments add much, and we don't have them below.
```suggestion
```python
from datashuttle import DataShuttle

project = DataShuttle("my_first_project")

created_folder_paths = project.create_folders(
    "sub-001", "ses-001_@DATE@", ["behav", "funcimg"]
)
```
```

in the session name acts as a wildcard. There are 
[many more options](https://datashuttle.neuroinformatics.dev/pages/how_tos/transfer-data.html#custom-transfers)
for customised transfer available.

Transfers can also be run directly in code through the Python API:
```python
from datashuttle import DataShuttle

project = DataShuttle("my_first_project")

project.transfer_custom(
    "rawdata", "all_sub", "ses-001_@*@", "behav"
)
```

## Logging

A complete record of all file transfers in a neuroscience project 
is invaluable, ensuring the full history of the project can be checked 
at any time. When a folder is created or data transferred in **datashuttle**, 
all details are logged to the local machine. Logs are stored
on the filesystem and can be viewed in a text editor or through
the graphical interface:

```{image} /_static/blog_images/datashuttle/logging-example-dark.png
:align: center
:class: only-dark
:width: 650px
```
```{image} /_static/blog_images/datashuttle/logging-example-light.png
:align: center
:class: only-light
:width: 650px
```
<br>

## Getting started with **datashuttle**

We have given a brief tour of **datashuttle**'s key features,

> but full details on getting started can be found on our 

Reviewer: ```suggestion
but full details on getting started can be found on the 
```
